syntax = "proto3";

package google.cloud.tools.snippetgen.snippetindex.v1;

option csharp_namespace = "Google.Cloud.Tools.SnippetGen.SnippetIndex.V1";

// The snippet index for a single client library.
message Index {
  // The Client Library these snippets are for.
  ClientLibrary client_library = 1;

  // The list of snippets.
  repeated Snippet snippets = 2;
}

// One sample.
// Parts of this information will be language specific.
message Snippet {
  // The region tag name. Does not include the square brackets or the START or
  // END indicators.
  string region_tag = 1;

  // The title of the snippet, for human consumption mostly. For generated
  // snippets this may be the snippet method or file name, or obtained from
  /// snippet configuration.
  string title = 2;

  // A description of the snippet, for human consumption mostly. For generated
  // snippets this may be the description of the service method, or obtained
  // from snippet configuration.
  string description = 3;

  // The file where the snippet code lives.
  // The path should be relative to the GitHub repo root and should not include
  // branch, tag, commitish, etc., as those will be the same as for the metadata
  // file.
  string file = 4;

  // The programming language name the snippet is written in.
  // This will match the client library language most of the time, but not
  // always. For instance, in .NET, libraries are written in C# but some samples
  // may be written in F# or VB .NET.
  // Note that this does not contain information about the snippet supported
  // platforms or language versions, etc. This is just a quick way to identify
  // the generally supported langauge.
  LanguageName language = 5;

  // The client library method this snippet is for.
  ClientMethod client_method = 6;

  // Wether this is the canonical snippet for the corresponding service method.
  // This is to be interpreted in conjunction with origin as follows:
  // For a given service method:
  // - A handwritten canonical takes precedence over
  // - A config canonical which in turn takes precedence over
  // - A baseline canonical.
  bool canonical = 7;

  // The origin of the snippet.
  Origin origin = 8;

  // The different segments of the snippet.
  // Must contain the FULL segment always.
  // There may be overlap between segments.
  repeated Segment segment = 9;

  // The origin of the snippet.
  enum Origin {
    // The origin has not been specified. Consumers should not see this value.
    ORIGIN_UNSPECIFIED = 0;

    // The snippet is generated from proto and client library information only,
    // no snippet configuration has been specified. This refers to SnippetGen
    // phase 1.
    BASELINE = 1;

    // The snippet is generated from a configuration. This refers to SnippetGen
    // phase 2.
    CONFIG = 2;

    // The snippet is handwritten.
    HANDWRITTEN = 3;
  }

  // A segment of the snippet.
  message Segment {
    // The line where this segment begins, inclusive.
    int32 start = 1;

    // The line where this segment ends, inclusive.
    int32 end = 2;

    // The type of the segment.
    SegmentType type = 3;

    // The type of the segment.
    // Basically describes what this segment shows.
    enum SegmentType {
      // The segment type has not been specified. Consumers should not see this
      // value.
      SEGMENT_TYPE_UNSPECIFIED = 0;

      // The full sample including import statements.
      // This corresponds to the sample as determined by the region tags.
      FULL = 1;

      // A shorter version of the full sample, may not include imports and some
      // langauge specific initialization code. This is to be used in contexts
      // in which the full aspects of the sample are made clear outside the
      // code.
      SHORT = 2;

      // The segment contains the service client initialization code only.
      // To be used in tutorials, codelabs, etc.
      CLIENT_INITIALIZATION = 3;

      // The segment contains the request initialization code only.
      // To be used in tutorials, codelabs, etc.
      REQUEST_INITIALIZATION = 4;

      // The segment contains the request execution code only.
      // To be used in tutorials, codelabs, etc.
      REQUEST_EXECUTION = 5;

      // The segment contains the response handling code only.
      // To be used in tutorials, codelabs, etc.
      RESPONSE_HANDLING = 6;
    }
  }
}

// A client library method.
// Will contain language specific information.
message ClientMethod {
  // The short name of the method, usually the name it is declared with.
  // This may not be unique within the service client because of overloads.
  string short_name = 1;

  // The fully qualified name of the method, which is the short_name qualified
  // by the full_name of the service client.
  // This value is redundant, but present to make it easier for consumers to
  // obtain it.
  // This may not be unique within the service client because of overloads.
  string full_name = 2;

  // Indicates wether this method is synchronous or asynchronous.
  // Some languages may support only one of the variants, in which case, this
  // field will always have the same value (for that language).
  bool async = 3;

  // Fully qualified type names of this method parameters in the same order as
  // they appear on the method declaration. Must be empty if the method has no
  // parameters.
  repeated string parameter_type = 4;

  // Fully qualified type name of this method result, if any.
  string result_type = 5;

  // The service client this method is declared in.
  ServiceClient client = 6;

  // The service method this client method is for.
  Method method = 7;
}

// A service client defined in the client library specified in Index.
// Will contain language specific information.
message ServiceClient {
  // The short name of the service client, usually the name it is declared with.
  // This may not be unique within the client library because of
  // namespaces/packages.
  string short_name = 1;

  // The fully qualified name of the service client, which is the short_name
  // qualified by the namespace/package/type name this client is declared in.
  // This will be unique within the client libray.
  string full_name = 2;
}

// A client library.
// Will contain language specific information.
message ClientLibrary {
  // The name of the client library. This value will be language dependent
  // and may or may not include the library version.
  // Usually this will be the name used to identify the library on per-language
  // package managers.
  // Examples: "Google.Cloud.Translate.V3",
  // "cloud.google.com/go/translate/apiv3".
  string name = 1;

  // The full version of the client library. May also be language dependent.
  // Cannot be updated on metadata generation, but on library release.
  // Examples: "4.3.0", "2.5.2-beta01"
  string version = 2;

  // The programming language name the library is written in.
  // Note that this does not contain information about the library supported
  // platforms or language versions, etc. This is just a quick way to identify
  // the generally supported langauge.
  LanguageName language = 3;

  // The API this client library is for.
  Api api = 4;
}

message Method {
  // The short name of the method, which is the name used to
  // declare it within the proto file. This is unique within the service,
  // but may not be unique within the API.
  string short_name = 1;

  // The full name of the method, which is the short name qualified
  // by the full name of the service in which it is declared.
  // This is globally unique.
  string full_name = 2;

  // The service this method is declared in.
  Service service = 3;
}

// A service defined in the API the client library referenced in Index is for.
message Service {
  // The short name of the service, which is the name used to
  // declare it within the proto file. This is usually, but not
  // absolutely necessarily, unique within an API.
  // Example: "TranslationService"
  string short_name = 1;

  // The full name of the service, which is the short name qualified
  // by the package of the proto in which it is declared.
  // This is globally unique.
  // Example: "google.cloud.translate.v3.TranslationService"
  string full_name = 2;
}

// An API
message Api {
  // The ID of the API, identical to the protobuf package
  // ending with a version number.
  // Example: "google.cloud.translate.v3"
  string id = 1;

  // The full version inferred from the end of the ID.
  // Examples: "v3", "v2beta1", "v1beta"
  string version = 2;
}

// A programming language
enum LanguageName {
  // The language has not been specified. Consumers should not see this value.
  LANGUAGE_UNSPECIFIED = 0;

  CPP = 1;

  C_SHARP = 2;

  DART = 3;

  F_SHARP = 4;

  GO = 5;

  JAVA = 6;

  NODE_JS = 7;

  PHP = 8;

  PYTHON = 9;

  RUBY = 10;

  SWIFT = 11;

  VB_NET = 12;
}